internal/backend/runtime/omni/controllers/omni/cluster_machine_teardown.go

```diff
@@ -0,0 +1,378 @@
+			},
+			{
+				Namespace: resources.DefaultNamespace,
+				Type:      omni.ClusterType,
+				Kind:      controller.InputQMappedDestroyReady,
+			},
+			{
+				Namespace: resources.DefaultNamespace,
+				Type:      omni.ClusterSecretsType,
+				Kind:      controller.InputQMapped,
+			},
+			{
+				Namespace: resources.DefaultNamespace,
+				Type:      omni.ClusterMachineIdentityType,
+				Kind:      controller.InputQMapped,
+			},
+			// TODO: drop after adding nodes + members audit
+			{
+				Namespace: resources.DefaultNamespace,
+				Type:      omni.MachineSetType,
+				Kind:      controller.InputQMapped,
+			},
+		},
+		Concurrency: optional.Some[uint](4),
+	}
+}
+
+// MapInput implements controller.QController interface.
+func (ctrl *ClusterMachineTeardownController) MapInput(ctx context.Context, logger *zap.Logger, r controller.QRuntime, ptr resource.Pointer) ([]resource.Pointer, error) {
+	switch ptr.Type() {
+	case omni.ClusterType:
+		mapper := mappers.MapByClusterLabel[*omni.Cluster, *omni.ClusterMachine]()
+
+		return mapper(ctx, logger, r, omni.NewCluster(ptr.Namespace(), ptr.ID()))
+	case omni.ClusterMachineIdentityType:
+		mapper := qtransform.MapperSameID[*omni.ClusterMachineIdentity, *omni.ClusterMachine]()
+
+		return mapper(ctx, logger, r, omni.NewClusterMachineIdentity(ptr.Namespace(), ptr.ID()))
+	case omni.ClusterSecretsType:
+		mapper := mappers.MapByClusterLabel[*omni.ClusterSecrets, *omni.ClusterMachine]()
+
+		return mapper(ctx, logger, r, omni.NewClusterSecrets(ptr.Namespace(), ptr.ID()))
+	case omni.MachineSetType:
+		machines, err := safe.ReaderListAll[*omni.ClusterMachine](ctx, r, state.WithLabelQuery(
+			resource.LabelEqual(omni.LabelMachineSet, ptr.ID()),
+		))
+		if err != nil {
+			return nil, err
+		}
+
+		return safe.Map(machines, func(r *omni.ClusterMachine) (resource.Pointer, error) {
+			return r.Metadata(), nil
+		})
+	}
+
+	return nil, fmt.Errorf("unexpected resource type %q", ptr.Type())
+}
+
+// Reconcile implements controller.QController interface.
+func (ctrl *ClusterMachineTeardownController) Reconcile(ctx context.Context, logger *zap.Logger, r controller.QRuntime, ptr resource.Pointer) error {
+	clusterMachine, err := safe.ReaderGetByID[*omni.ClusterMachine](ctx, r, ptr.ID())
+	if err != nil {
+		if state.IsNotFoundError(err) {
+			return nil
+		}
+
+		return err
+	}
+
+	clusterName, ok := clusterMachine.Metadata().Labels().Get(omni.LabelCluster)
+	if !ok {
+		return fmt.Errorf("failed to determine the cluster name of the cluster machine %q", ptr.ID())
+	}
+
+	logger = logger.With(zap.String("machine", clusterMachine.Metadata().ID()), zap.String("cluster", clusterName))
+
+	cluster, err := safe.ReaderGetByID[*omni.Cluster](ctx, r, clusterName)
+	if err != nil && !state.IsNotFoundError(err) {
+		return err
+	}
+
+	// remove finalizer without any actions if the cluster resource is either absent or is tearing down
+	if cluster == nil || cluster.Metadata().Phase() == resource.PhaseTearingDown {
```

Comment:
probable moved down to the line 163 (as it makes sense only if the phase is tearing down and the finalizer is set